src/crab/type_domain.hpp

```diff
@@ -0,0 +1,195 @@
+// Copyright (c) Prevail Verifier contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include <unordered_map>
+
+#include "crab/abstract_domain.hpp"
+#include "crab/cfg.hpp"
+#include "linear_constraint.hpp"
+#include "string_constraints.hpp"
+
+namespace crab {
+
+enum class region {
+	T_CTX,
+	T_STACK,
+	T_PACKET,
+	T_SHARED
+};
+
+struct ptr_no_off_t {
+    region r;
+
+    ptr_no_off_t() = default;
+    ptr_no_off_t(const ptr_no_off_t &) = default;
+    ptr_no_off_t(ptr_no_off_t &&) = default;
+    ptr_no_off_t &operator=(const ptr_no_off_t &) = default;
+    ptr_no_off_t &operator=(ptr_no_off_t &&) = default;
+    ptr_no_off_t(region _r) : r(_r) {}
+
+    friend std::ostream& operator<<(std::ostream& o, const ptr_no_off_t& p);
+    friend bool operator==(const ptr_no_off_t& p1, const ptr_no_off_t& p2);
+    friend bool operator!=(const ptr_no_off_t& p1, const ptr_no_off_t& p2);
+};
+
+struct ptr_with_off_t {
+    region r;
+    int offset;
+
+    ptr_with_off_t() = default;
+    ptr_with_off_t(const ptr_with_off_t &) = default;
+    ptr_with_off_t(ptr_with_off_t &&) = default;
+    ptr_with_off_t &operator=(const ptr_with_off_t &) = default;
+    ptr_with_off_t &operator=(ptr_with_off_t &&) = default;
+    ptr_with_off_t(region _r, int _off) : r(_r), offset(_off) {}
+
+    friend std::ostream& operator<<(std::ostream& o, const ptr_with_off_t& p);
+    friend bool operator==(const ptr_with_off_t& p1, const ptr_with_off_t& p2);
```

Comment:
Why `operator==` is a friend. It shouldn't. Probably you want it so it's easy to use it as a key in a map? You can customize the key in maps so no need for friends.

Reply:
And I want `r` and `offset` to be private

Reply:
Not for map keys. `ptr_t` (pointer type) is a variant consisting of `ptr_with_off_t` and `ptr_no_off_t`. When comparing two variants, we need to define equality operators. Hence, I have the `operator==`. However, I tried to remove the `operator==` by specializing the function `equal_to` for `ptr_t` but I couldn't make it work.